File: app/src/main/java/com/seungjun/randomox/view/JoinPopup.java
package com.seungjun.randomox.view;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.preference.PreferenceManager;
import android.text.TextUtils;
import android.view.View;
import android.view.Window;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.seungjun.randomox.R;
import com.seungjun.randomox.network.RetrofitApiCallback;
import com.seungjun.randomox.network.RetrofitClient;
import com.seungjun.randomox.network.data.HeaderInfo;
import com.seungjun.randomox.network.data.UserInfo;
import com.seungjun.randomox.utils.CommonUtils;
import com.seungjun.randomox.utils.PreferenceUtils;
import com.wang.avi.AVLoadingIndicatorView;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class JoinPopup extends Dialog {

    private Context context;

    @BindView(R.id.input_nickname)
    EditText inputNickName;

    @BindView(R.id.input_pw)
    EditText inputPw;

    @BindView(R.id.error_text)
    TextView errorText;

    @BindView(R.id.join_progress)
    AVLoadingIndicatorView joinProgress;

    @BindView(R.id.btn_join)
    TextView btnJoin;


    private RetrofitClient networkClient;


    public JoinPopup(Context context) {

        super(context, android.R.style.Theme_Translucent_NoTitleBar);

        requestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.view_join_dialog);

        this.context = context;

        ButterKnife.bind(this);

        networkClient = RetrofitClient.getInstance(context).createBaseApi();
    }

    @OnClick(R.id.btn_join)
    public void clickJoin(){

        String nickname = inputNickName.getText().toString();
        String password = inputPw.getText().toString();

        if(TextUtils.isEmpty(nickname)){
            errorText.setVisibility(View.VISIBLE);
            errorText.setText(context.getResources().getString(R.string.error_empty_name));

            return;
        }


        if(!CommonUtils.isValidName(nickname)){
            errorText.setVisibility(View.VISIBLE);
            errorText.setText(context.getResources().getString(R.string.error_nonok_name));

            return;
        }


        if (TextUtils.isEmpty(password)){
            errorText.setVisibility(View.VISIBLE);
            errorText.setText(context.getResources().getString(R.string.error_empty_pw));

            return;
        }


        if(!CommonUtils.isValidPw(password)){
            errorText.setVisibility(View.VISIBLE);
            errorText.setText(context.getResources().getString(R.string.error_nonok_pw));

            return;
        }


        inputNickName.setEnabled(false);
        inputNickName.setFocusable(false);

        inputPw.setEnabled(false);
        inputPw.setFocusable(false);

        btnJoin.setVisibility(View.GONE);
        joinProgress.setVisibility(View.VISIBLE);
        this.setCancelable(false);


        networkClient.callPostJoin(new RetrofitApiCallback() {
            @Override
            public void onError(Throwable t) {

                inputNickName.setEnabled(true);
                inputNickName.setFocusable(true);

                inputPw.setEnabled(true);
                inputPw.setFocusable(true);

                btnJoin.setVisibility(View.VISIBLE);
                joinProgress.setVisibility(View.GONE);
                JoinPopup.this.setCancelable(true);

                errorText.setText(context.getResources().getString(R.string.error_network_unkonw));
            }

            @Override
            public void onSuccess(int code, Object resultData) {

                HeaderInfo result = (HeaderInfo)resultData;

                if(result.reqCode == 0){
                    Toast.makeText(context, "가입 성공! 로그인 해주세요!", Toast.LENGTH_SHORT).show();

                    JoinPopup.this.dismiss();
                }else{

                    inputNickName.setEnabled(true);
                    inputNickName.setFocusable(true);

                    inputPw.setEnabled(true);
                    inputPw.setFocusable(true);

                    btnJoin.setVisibility(View.VISIBLE);
                    joinProgress.setVisibility(View.GONE);
                    JoinPopup.this.setCancelable(true);

                    errorText.setText(result.reqMsg);
                }
            }

            @Override
            public void onFailed(int code) {


                inputNickName.setEnabled(true);
                inputNickName.setFocusable(true);

                inputPw.setEnabled(true);
                inputPw.setFocusable(true);

                btnJoin.setVisibility(View.VISIBLE);
                joinProgress.setVisibility(View.GONE);
                JoinPopup.this.setCancelable(true);

            }
        }, nickname, password, PreferenceUtils.getInstance(context).getUserFcmKey());
    }


    @Override
    public void show() {

        if(((Activity)context).isFinishing())
            return;

        super.show();
    }

    @Override
    public void dismiss() {
        super.dismiss();
    }
}

File: app/src/main/java/com/seungjun/randomox/view/JoinPopup.kt
package com.seungjun.randomox.view

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.text.TextUtils
import android.view.View
import android.view.Window
import android.widget.Toast
import com.seungjun.randomox.R
import com.seungjun.randomox.network.RetrofitApiCallback
import com.seungjun.randomox.network.RetrofitClient
import com.seungjun.randomox.network.data.HeaderInfo
import com.seungjun.randomox.utils.CommonUtils
import com.seungjun.randomox.utils.PreferenceUtils
import kotlinx.android.synthetic.main.view_join_dialog.*

class JoinPopup(context: Context) : Dialog(context, android.R.style.Theme_Translucent_NoTitleBar) {

    var mContext: Context? = null

    init {

        requestWindowFeature(Window.FEATURE_NO_TITLE)
        setContentView(R.layout.view_join_dialog)

        btn_join.setOnClickListener {
            clickJoin()
        }

        mContext = context
    }


    fun clickJoin() {

        val nickname = input_nickname.text.toString()
        val password = input_pw.text.toString()

        if (TextUtils.isEmpty(nickname)) {
            error_text.apply {
                visibility = View.VISIBLE
                text = context.resources.getString(R.string.error_empty_name)
            }
            return
        }


        if (!CommonUtils.isValidName(nickname) || nickname.length < 2 || nickname.length > 12) {
            error_text.apply {
                visibility = View.VISIBLE
                text = context.resources.getString(R.string.error_nonok_name)
            }
            return
        }


        if (TextUtils.isEmpty(password)) {
            error_text.apply {
                visibility = View.VISIBLE
                text = context.resources.getString(R.string.error_empty_pw)
            }
            return
        }


        if (!CommonUtils.isValidPw(password)) {
            error_text.apply {
                visibility = View.VISIBLE
                text = context.resources.getString(R.string.error_nonok_pw)
            }
            return
        }

        input_nickname.apply {
            isEnabled = false
            isFocusable = false
        }

        input_pw.apply {
            isEnabled = false
            isFocusable = false
        }

        btn_join.visibility = View.GONE
        join_progress.visibility = View.VISIBLE

        this.setCancelable(false)


        RetrofitClient.callPostJoin(object : RetrofitApiCallback<HeaderInfo> {
            override fun onError(t: Throwable) {

                failedJoin(context.resources.getString(R.string.error_network_unkonw))
            }

            override fun onSuccess(code: Int, resultData: HeaderInfo) {

                if (resultData.reqCode == 0) {
                    Toast.makeText(context, "가입 성공! 로그인 해주세요!", Toast.LENGTH_SHORT).show()

                    this@JoinPopup.dismiss()
                } else {
                    failedJoin(resultData.reqMsg)
                }
            }

            override fun onFailed(code: Int) {

                failedJoin(context.resources.getString(R.string.error_network_unkonw))

            }
        }, nickname, CommonUtils.getAES256(context, password), PreferenceUtils.getInstance(context).userFcmKey!!)
    }


    fun failedJoin(msg: String) {

        input_nickname.apply{
            isEnabled = true
            isFocusableInTouchMode = true
            isFocusable = true
        }

        input_pw.apply{
            isEnabled = true
            isFocusableInTouchMode = true
            isFocusable = true
        }

        btn_join.visibility = View.VISIBLE
        join_progress.visibility = View.GONE
        this@JoinPopup.setCancelable(true)

        error_text.apply {
            visibility = View.VISIBLE
            text = msg
        }
    }


    override fun show() {

        if ((mContext as Activity).isFinishing)
            return

        super.show()
    }

    override fun dismiss() {
        super.dismiss()
    }
}
